### ISSUE TITLE
Improper constraint check failure with new "Any" type and Std.instance

### ISSUE THREAD

**Issue by SmerkyG** [HaxeFoundation/haxe#6649]
Easily reproducible constraint check problem with trying to get a Std.instance of something that's Any. It works fine if you use Dynamic instead of Any in Haxe 3.4.0:
```haxe
class Test {
    var any:Any;
    function foo() {
        var test = Std.instance(any, Test);
	}
}
```
```
Test.hx:4: characters 19-31 : Constraint check failure for instance.S
Test.hx:4: characters 19-31 : Test should be Any
Test.hx:4: characters 19-31 : Constraint check failure for instance.T
Test.hx:4: characters 19-31 : Any should be { }

**Comment by SmerkyG** [comment]
Additionally, there appear to be some unification issues between Class<Any> and Class<Dynamic> that come up when calling Type.getClass(a) where a:Any. It seems that there are several such edge cases regarding convertability between Any and Dynamic when they Any is used as a generic type parameter for standard library functions.

**Comment by RealyUniqueName** [comment]
Changing this to 4.1 as it is expected to affect new asys API. 
See https://github.com/HaxeFoundation/haxe/pull/9005#issue-348366685

**Comment by RealyUniqueName** [comment]
Not related to asys API anymore.

Changing `Any` to behave like a wildcard type parameter (as `Dynamic`) alone won't solve the issue.
In case of `Dynamic` it is not propagated, while `Any` is propagated.  And `Std.downcast` infers the type of the first argument.
And instead of changing the behavior of Haxe type system std (or any other lib) could have a method for such cases:
```haxe
static public function as<T,S>(value:T, c:Class<S>):Null<S> {
  return Std.isOfType(value, c) ? cast value : null;
}
```

**Comment by RealyUniqueName** [comment]
Fixed in #9741
